# Testing

We are implementing a growing number of tests to ensure that the outputs of the pipeline are correct.

## Data

We have 6 test datasets currently:
- Equal reads simulation: art_illumina generating a metagenome where each organism yields the same number of reads
- Equal coverage simulation: art_illumina generating a metagenome where each organism is given the same coverage depth
- a host-depleted subset of sample 473
- SRR18369973
- SRR21986403
- SRR19479275

The SRA datasets can be downloaded by running `.test/getdata.sh`; this requires singularity.

The simulated datasets can be generated by running
```
snakemake --snakefile .test/Snakefile --directory .test/simulated/
```

## Testing pipeline execution
Pipeline runs can be executed by running the `test.sh` script as follows:
```
bash test.sh <stage> <dataset>
# eg
bash test.sh preproess equalreads
```

Results will be written to `tmp<stage>_<dataset>/`.

## Testing pipeline results
A number of pipeline results are required for the full test suite. As a shortcut to generating all these, run
```
bash integration.sh
```
which will execute all the pipeline runs needed by the test suite.

Tests are implemented in pytest.  Run `pytest` in the root of the repository.
